src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/mode-base-server/mode-base-server.h>
+#include <app/clusters/on-off-server/on-off-server.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <platform/DiagnosticDataProvider.h>
+
+// using namespace std;
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using chip::Protocols::InteractionModel::Status;
+using BootReasonType = GeneralDiagnostics::BootReasonEnum;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeBase {
+
+std::map<uint32_t, Instance *> Instance::ModeBaseAliasesInstanceMap;
+
+bool Instance::HasFeature(Feature feature) const
+{
+    return (mFeature & to_underlying(feature)) != 0;
+}
+
+bool Instance::isAliasCluster() const
+{
+    for (unsigned int AliasedCluster : AliasedClusters)
+    {
+        if (mClusterId == AliasedCluster)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+CHIP_ERROR Instance::Init()
+{
+    // Check that the cluster ID given is a valid mode base alias cluster ID.
+    if (!isAliasCluster())
+    {
+        ChipLogError(Zcl, "ModeBase: The cluster with ID %lu is not a mode base alias.", long(mClusterId));
+        return CHIP_ERROR_INVALID_ARGUMENT; // todo is this the correct error?
+    }
+
+    // Check if the cluster has been selected in zap
+    if (!emberAfContainsServer(mEndpointId, mClusterId))
+    {
+        ChipLogError(Zcl, "ModeBase: The cluster with ID %lu was not enabled in zap.", long(mClusterId));
+        return CHIP_ERROR_INVALID_ARGUMENT; // todo is this the correct error?
+    }
+
+    ReturnErrorOnFailure(chip::app::InteractionModelEngine::GetInstance()->RegisterCommandHandler(this));
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(AppInit());
+
+    ModeBaseAliasesInstanceMap[mClusterId] = this;
+
+    // StartUp behavior relies on CurrentMode StartUpMode attributes being non-volatile.
+    if (!emberAfIsKnownVolatileAttribute(mEndpointId, mClusterId, Attributes::CurrentMode::Id) &&
+        !emberAfIsKnownVolatileAttribute(mEndpointId, mClusterId, Attributes::StartUpMode::Id))
+    {
+        // Read the StartUpMode attribute and set the CurrentMode attribute
+        // The StartUpMode attribute SHALL define the desired startup behavior of a
+        // device when it is supplied with power and this state SHALL be
+        // reflected in the CurrentMode attribute.  The values of the StartUpMode
+        // attribute are listed below.
+
+        DataModel::Nullable<uint8_t> startUpMode = GetStartUpMode();
+        if (!startUpMode.IsNull())
+        {
+#ifdef EMBER_AF_PLUGIN_ON_OFF
```

Comment:
This is extremely unfortunate, and makes me sad, but such is life right now. You cannot work around this for now, but we have to eventually find a way. This does not scale.

Reply:
Why can't we remove it? I left it in because of the way it was previously written. But we are still checking if an onoff clusters exists on the same endpoint. So if the onoff endpoint is not used, all we do is call an extra if.

Reply:
I have removed these. We are anyway checking for the presence of the OnOff cluster and none of the functions used in these checks will be missing if the OnOff cluster is not used.